connect/runtime/src/test/java/org/apache/kafka/connect/runtime/isolation/MultiVersionTest.java

```diff
@@ -0,0 +1,296 @@
+        Map<Path, List<VersionedPluginBuilder.BuildInfo>> artifacts = new HashMap<>();
+        for (String v : versions) {
+            for (VersionedPluginBuilder.VersionedTestPlugin pluginType: pluginTypes) {
+                VersionedPluginBuilder builder = new VersionedPluginBuilder();
+                builder.include(pluginType, v);
+                artifacts.put(builder.build(pluginType + "-" + v), builder.buildInfos());
+            }
+        }
+        return artifacts;
+    }
+
+    public static final String DEFAULT_ISOLATED_ARTIFACTS_LATEST_VERSION;
+    public static final Map<Path, List<VersionedPluginBuilder.BuildInfo>> DEFAULT_ISOLATED_ARTIFACTS;
+    public static final Map<Path, List<VersionedPluginBuilder.BuildInfo>> DEFAULT_COMBINED_ARTIFACT;
+    public static final Plugins MULTI_VERSION_PLUGINS;
+    public static final Map<VersionedPluginBuilder.VersionedTestPlugin, String> DEFAULT_COMBINED_ARTIFACT_VERSIONS;
+
+    static {
+
+        String[] defaultIsolatedArtifactsVersions = new String[]{"1.1.0", "2.3.0", "4.3.0"};
+        try {
+            DEFAULT_ISOLATED_ARTIFACTS = buildIsolatedArtifacts(
+                defaultIsolatedArtifactsVersions, VersionedPluginBuilder.VersionedTestPlugin.values()
+            );
+            DEFAULT_ISOLATED_ARTIFACTS_LATEST_VERSION = "4.3.0";
+            DEFAULT_COMBINED_ARTIFACT_VERSIONS = new HashMap<>();
+
+            VersionedPluginBuilder builder = new VersionedPluginBuilder();
+            builder.include(VersionedPluginBuilder.VersionedTestPlugin.SOURCE_CONNECTOR,
+                DEFAULT_COMBINED_ARTIFACT_VERSIONS.computeIfAbsent(VersionedPluginBuilder.VersionedTestPlugin.SOURCE_CONNECTOR, k -> "0.0.0"));
+            builder.include(VersionedPluginBuilder.VersionedTestPlugin.SINK_CONNECTOR,
+                DEFAULT_COMBINED_ARTIFACT_VERSIONS.computeIfAbsent(VersionedPluginBuilder.VersionedTestPlugin.SINK_CONNECTOR, k -> "0.1.0"));
+            builder.include(VersionedPluginBuilder.VersionedTestPlugin.CONVERTER,
+                DEFAULT_COMBINED_ARTIFACT_VERSIONS.computeIfAbsent(VersionedPluginBuilder.VersionedTestPlugin.CONVERTER, k -> "0.2.0"));
+            builder.include(VersionedPluginBuilder.VersionedTestPlugin.HEADER_CONVERTER,
+                DEFAULT_COMBINED_ARTIFACT_VERSIONS.computeIfAbsent(VersionedPluginBuilder.VersionedTestPlugin.HEADER_CONVERTER, k -> "0.3.0"));
+            builder.include(VersionedPluginBuilder.VersionedTestPlugin.TRANSFORMATION,
+                DEFAULT_COMBINED_ARTIFACT_VERSIONS.computeIfAbsent(VersionedPluginBuilder.VersionedTestPlugin.TRANSFORMATION, k -> "0.4.0"));
+            builder.include(VersionedPluginBuilder.VersionedTestPlugin.PREDICATE,
+                DEFAULT_COMBINED_ARTIFACT_VERSIONS.computeIfAbsent(VersionedPluginBuilder.VersionedTestPlugin.PREDICATE, k -> "0.5.0"));
+            DEFAULT_COMBINED_ARTIFACT = Collections.singletonMap(builder.build("all_versioned_artifact"), builder.buildInfos());
```

Comment:
nit: Why is this computeIfAbsent, when we are constructing the map for the first time?

Reply:
So `computeIfAbsent` conveniently returns the computed value, which is used in the builder. A simple `put` would return the previous value which would be `null`. We can separate out this into individual `put` and 'get` calls if this looks a bit hard to read.